shared/binarySearch.go

```diff
@@ -0,0 +1,104 @@
+package shared
+
+import (
+	"strings"
+)
+
+/*Performs binary search on
+First we identify a match. When we the match, we branch out and find upper and lower bound.*/
+func BinarySearch(genome string, read string, suffixArray []int) (int, int) {
+
+	//add sentinel if not present
+	var sb strings.Builder
+	if genome[len(genome)-1] != '$' {
+		sb.WriteString(genome)
+		sb.WriteByte('$')
+		genome = sb.String()
+	}
+
+	//edgecase where read is empty or longer than genome. Return empty interval (genome[0:0[)
+	if read == "" || len(read) > len(genome) {
+		return 0, 0
+	}
+	upperBound := upperBound(suffixArray, genome, read)
+	lowerBound := lowerBound(suffixArray, genome, read)
+	if upperBound != lowerBound {
+		return lowerBound, upperBound
+	}
+	return 0, 0
```

Comment:
Why is it important to return `0, 0` if there are no matches? Any empty interval would do, I should think, so you can always return `lowerBound, upperBound`, right?